plugin/utils/shell.py

```diff
@@ -86,6 +87,183 @@ def create_lods():
+            
+            #Debug print
+            print("-------------------------------------------------------------")
+            print("Posed bone")
+            print("name: "+str(i.name))
+            print("parent: "+str(i.parent.name))
+            
+            #Get pose mode position of bonebase
+            bonebase_pos_pose = Vector(i.matrix.translation)
+            
+            #Reset position
+            i.location = Vector((0, 0, 0))
+            #Update location after moving.
+            bpy.context.view_layer.update()
+            
+            #Get rest mode position of bonebase
+            bonebase_pos_rest = Vector(i.matrix.translation)
+            
+            #Calculate delta offset
+            bonebase_pos_delt = Vector(bonebase_pos_pose - bonebase_pos_rest)
+            
+            #Set back to original position
+            i.matrix.translation = Vector(bonebase_pos_pose)
+            
+            #Print locations
+            print("bonebase rest position: "+str(bonebase_pos_rest))
+            print("bonebase pose position: "+str(bonebase_pos_pose))
+            print("bonebase delt position: "+str(bonebase_pos_delt))
+
+            
+            #Creating the nodes            
+            #We switch to edit mode to use edit bones, as they do not exist outside of edit mode.
+            #Switch to edit mode
+            bpy.ops.object.mode_set(mode='EDIT')
+            
+            #Get edit bones
+            bonebase = armature.data.edit_bones.get(i.name)
+            boneparent = armature.data.edit_bones.get(i.parent.name)
+            
+            #Creating the node bone
+            bonenode = armature.data.edit_bones.new("NODE_"+str(i.name))
+            #NOTE: Maybe here we set the bonenode to boneparent if it starts with NODE_ already
+            
+            #Copy parent length to node
+            bonenode.length = boneparent.length
+            #Copy parent matrix to node
+            bonenode.matrix = boneparent.matrix.copy()
+            
+            #Set parent of bonenode to boneparent
+            bonenode.parent = boneparent
+            #Set parent of bonebase to bonenode
+            bonebase.parent = bonenode
+            #Node creation complete
+            
+            #Switch to pose mode.
+            bpy.ops.object.mode_set(mode='POSE')
+            
+            
+            
+            #------------------------------------
+            
+            #OFFSET NODE LOCATION
+            #Applying offsets to bonenode
+            #Switch to pose mode
+            bpy.ops.object.mode_set(mode='POSE')
+            
+            #Clear location
+            i.location = Vector((0,0,0))
+            #Clear rotation
+            i.rotation_quaternion = Quaternion((1,0,0,0))
+            
+            #Offset bonenode location
+            i.parent.matrix.translation = i.parent.matrix.translation + bonebase_pos_delt
```

Comment:
let's look at the matrix math later lol

Reply:
for context - the plugin itself actually has to use a hack to try and get valid bone roll for all bones. blender has a big numerical precision issue there, even if the math is perfectly correct.
so we have to verify if your math is correct or not